pennylane/qnn.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the :class:`~.KerasLayer` class for integrating QNodes with the Keras
+layer API."""
+import functools
+import inspect
+from collections.abc import Iterable
+from typing import Optional
+
+try:
+    import tensorflow as tf
+    from tensorflow.keras.layers import Layer
+    from pennylane.interfaces.tf import to_tf
+
+    CORRECT_TF_VERSION = int(tf.__version__.split(".")[0]) > 1
+except ImportError:
+    from abc import ABC
+
+    Layer = ABC
+    CORRECT_TF_VERSION = False
+
+from pennylane.qnodes import QNode
+
+INPUT_ARG = "inputs"
+
+
+class KerasLayer(Layer):
+    """A Keras layer for integrating PennyLane QNodes with the Keras API.
+
+    This class converts a :class:`~.QNode` to a Keras layer. The QNode must have a signature that
+    satisfies the following conditions:
+
+    - Contain an ``inputs`` argument for inputting data. All other arguments are treated as
+      weights within the QNode.
+    - All arguments must accept an array or tensor, e.g., arguments should not use nested lists
+      of different lengths.
+    - All arguments, except ``inputs``, must have no default value.
+    - The ``inputs`` argument is permitted to have a default value provided the gradient with
+      respect to ``inputs`` is not required.
+    - There cannot be a variable number of positional or keyword arguments, e.g., no ``*args`` or
+      ``**kwargs`` present in the signature.
+
+    The QNode weights are initialized within the :class:`~.KerasLayer`. Upon instantiation,
+    a ``weight_shapes`` dictionary must hence be passed which describes the shapes of all
+    weights in the QNode. The optional ``weight_specs`` argument allows for a more fine-grained
+    specification of the QNode weights, such as the method of initialization and any
+    regularization or constraints. If not specified, weights will be added using the Keras
+    default initialization and without any regularization
+    or constraints.
+
+    **Example usage:**
+
+    .. code-block:: python
```

Comment:
It might be useful having a sentence or two above this code-block, just introducing the example.